Fix EPQ for DML operations

HustonMmmavr opened the pull request:
## Problem description

Currently `EPQ` mechanism doesnt't work correctly at `ORCA` planner. Here is example of query which proves that:

```sql
-- gp_enable_global_deadlock_detector is on 
create table test as select 0 as i distributed randomly;

-- Start update transaction
1: begin;
1: update test set i = i + 1;

-- Second session tries to delete modified row and
-- it waits for result of update
2&: delete from test where i = 0;
1: end;
-- as a result updated row was deleted - this is
-- not vaild behaviour (there must be no deletion)
DELETE 1
```

Unlike postgres plan, `SEQSCAN` node (corresponding to `ModifyTable` node) doesn't points to the target(result) relation of the plan. `scanrelid` of `SEQSCAN` node points to the `RTE` `2`, but `resultRelation` has index `1`. Here is a plan:

```
{PLANNEDSTMT 
    ...
   :planTree 
      {MODIFYTABLE
      ... 
      :resultRelations (i 1)
      ...
      :plans (
         {SEQSCAN 
         ...
         :plan_node_id 1
         ... 
         :scanrelid 2
         ...
         }
      )
      ...
      }
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 8 
      ...
      }
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 2 
      ...
      }
   )
   :resultRelations (i 1)
   :relationOids (o 16548 16548 16548)
   ...
   }
```

EPQ mechanism through function call chain calls `ExecScanFetch`.
Due to incorrect `scanrelid` at `SEQSCAN` node next branch under  `if (estate->es_epqTupleSet[scanrelid - 1])` won't be executed. Thus instead of working with slot from `es_epqTuple` cache - slot with an old row is returned from access method routine - `(*accessMtd) (node)` (which uses snapshot for delete operation). As a result `EvalPlanQual` returns not null slot with an old version of row to `ExecDelete`, but [tupleid points to the updated version and deletion is performed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/executor/nodeModifyTable.c#L943-L957)

```c
static inline TupleTableSlot *
ExecScanFetch(ScanState *node,
			  ExecScanAccessMtd accessMtd,
			  ExecScanRecheckMtd recheckMtd)
{
	EState	   *estate = node->ps.state;
    //...
	if (estate->es_epqTupleSlot != NULL)
	{
        // ...
		Index		scanrelid = ((Scan *) node->ps.plan)->scanrelid;

        // ...
		else if (estate->es_epqTupleSlot[scanrelid - 1] != NULL)
		{
			TupleTableSlot *slot = node->ss_ScanTupleSlot;

			/* Return empty slot if we already returned a tuple */
			if (estate->es_epqScanDone[scanrelid - 1])
				return ExecClearTuple(slot);
			/* Else mark to remember that we shouldn't return more */
			estate->es_epqScanDone[scanrelid - 1] = true

			slot = estate->es_epqTupleSlot[scanrelid - 1];
			//...
			return slot;
		}
	}

	/*
	 * Run the node-type-specific access method function to get the next tuple
	 */
	return (*accessMtd) (node);
}
```

At the initial Query structure (parsed and analyzed query) we have right links to `RTE`'s, so to fix the problem with `scanrelid` - we want to transfer this information to the initial algebrized plan tree of orca. So we assign the `queryid` to each query structure (for subqueruies `queryid` assigned too) and, also, we assign `queryid` to table descriptors `CDXLTableDescr` which explicitly refers to target relation of DML (`ModifyTable`). After orca planning transformations this value exposes to which target relation of DML (`ModifyTable`) operation is referred the current table descriptor. And this allows to correctly assign `RTE` indexes in result plan under transformation from `DXL` representation.

## Forced changes

After solving problem with `scanrelid` for `*SCAN` (`SEQSCAN`, `INDEXSCAN` and others) nodes of DML (`ModifyTable`) node
(`scanrelid` of `*SCAN` node corresponding to `ModifyTable` node points to the result/target relation of operation) output of `EXPLAIN` has changed. Thus `regress` tests and `modify_table_data_corrupt` test from `isolation2` was changed.
These changes could be divided into two groups:

### Tests that contains one table at DML (`ModifyTable`) operation

Here is exapmle of some test cases (links from master):

* [updatable_views_optimizer - prefixes at coulumn names and table alias are not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/updatable_views_optimizer.out#L404-L416)

* [update_gp_optimizer - prefixes at coulumn names and table alias are not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/update_gp_optimizer.out#L435-L443)

* [bfv_dml_optimizer - alias for table name is not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/bfv_dml_optimizer.out#L185-L189)

At these tests DML (`ModifyTable`) operation is performed on the one table. Due to fixes of this patch, `ModifyTable` operations at these tests has only one (single) RTE - target relation, unlike it was before (scanrelid of `*SCAN` plan node (correspoinmg to `ModifyTalbe`) didn't point to the result (target) relation and plan contained two rtables, but there should be only one). Thus prefixes to column names at queries, that uses one table is not shown, like it is in postgres. Here is some links to code, related to showing prefix ([setting flag to show prefix](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain_gp.c#L2109) and and function `get_variable`, where prefix [may be appended](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/utils/adt/ruleutils.c#L6852)).
Also at some tests `explain` aliases was [deleted, because them are equal to the original relation name](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L4033-L4035)

### Tests with explain (used more than one relation)

Here is an example of some test case (links from master):

* [qp_subquery_optimizer - aliases for table name in explain shouldn't have "_1" suffix](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/qp_subquery_optimizer.out#L1446-L1453)

This test group was changed due to fixes of `scanrelid` too. `scanrelid` for `*SCAN` node of corresponoding `ModifyTable` node now points to the target(result) relation (and rtable list doesn't contain duplicate of `ModifyTable` target relation). Thus `ExplainPrintPlan` calculates list of `relation` names for explain a bit different (in comparsion to what was before) (calcultion is [based on "rels_used" bitmap](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L895-L897), which now [calculates correctly](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L1338-L1383) - `ModifyTable` node doesn't insert at bitmap by the fact unused rti, like it was before, so there is no phantom relation name in `rtable_names`). So position of table names (`explain` aliases) with suffixes changed due to fixes of `scanrelid`'. Like at previous test group at some tests `explain` aliases was [deleted, because them are equal to the original relation name](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L4033-L4035)

### Tests for the patch

The `gdd/concurrent_update.sql` (`expected/gdd/concurrent_update_optimizer.out`) was partially [ported from the 6X_STABLE](https://github.com/greenplum-db/gpdb/blob/f1bf39d23057f24aac1189577262340e4b72054b/src/test/isolation2/sql/gdd/concurrent_update.sql) branch and extended with test case from the description.

HustonMmmavr commented:
This patch should be backported to 6X_STABLE, I've already created temp branch, based on this patch (there were many conflicts). I'll open PR to 6X_STABLE after applying this patch, because changes may be required after review.

THANATOSLAVA commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -723,6 +725,33 @@ CTranslatorQueryToDXL::TranslateQueryToDXL()
+		assigned_queryid = m_query_id;
+	}
+
+	CDXLTableDescr *table_descr = CTranslatorUtils::GetTableDescr(
```
Looks like `CTranslatorUtils::GetTableDescr()` is only used in this file. Can we just rewrite that one instead of adding this wrapper?

THANATOSLAVA commented:
>Unlike postgres plan, SEQSCAN node (corresponding to ModifyTable node) doesn't points to the target(result) relation of the plan. scanrelid of SEQSCAN node points to the RTE 2, but resultRelation has index 1.

In the planned statement that you posted, the first `RTE` has `requiredPerms = 8`, meaning it grants DELETE right.`PLANNEDSTMT->resultRelations` points to `RTE 1`, meaning the results of deletion goes to `RTE 1`. So far this makes sense.

Now `MODIFYTABLE->SEQSCAN->scanrelid` points to `RTE 2`. And the second `RTE` has `requiredPerms = 2`, meaning it grants SELECT right. So what is `RTE 2` for? I wonder if `RTE 2` should be appended to the range table list in the first place.

Another question is, we should have an `RTE` that grants UPDATE right. Is that one missing or just not pasted there? Did the update `i += 1` actually happen?

THANATOSLAVA commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -536,6 +536,59 @@ CTranslatorDXLToPlStmt::SetParamIds(Plan *plan)
+	BOOL is_existed_rte = false;
+	if (query_id_of_target_descr != UNASSIGNED_QUERYID)
+	{
+		ULONG *key = GPOS_NEW(m_mp) ULONG(query_id_of_target_descr);
```
We should also move the `index` definition on line 558 to inside the `else` block. It won't be used in case of `usedIndex`.

THANATOSLAVA commented:
A general question. What's the relation between `queryID` under the planned statement node and `assigned_query_id` introduced by this PR?

```
DETAIL:     {PLANNEDSTMT
   :commandType 2
   :planGen 1
   :queryId 0
```

HustonMmmavr commented:
> In the planned statement posted here, the first RTE has requiredPerms = 8, meaning it grants DELETE right. PLANNEDSTMT->resultRelations points to RTE 1, meaning the results of deletion goes to RTE 1. So far this makes sense.

> Now MODIFYTABLE->SEQSCAN->scanrelid points to RTE 2. And the second RTE has requiredPerms = 2, meaning it grants SELECT right. So what is RTE 2 for? I wonder if RTE 2 should be appended to the range table list in the first place. Instead, RTE 1->requiredPerms should be a bitwise OR of the SELECT and DELETE rights.

Here is a final planned statement after this patch (from orca optimizer)
```log
{PLANNEDSTMT 
   :commandType 4 
   :queryId 0 
   ...
   :planTree 
      {MODIFYTABLE 
      ...
      :resultRelations (i 1)
      ...
      :plans (
         {SEQSCAN 
         ...
         :plan_node_id 1 
         ... 
         :scanrelid 1
         ...
         }
      )
      ...
      }
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16388 
      ...
      :requiredPerms 8 
      ...
      }
   )
   :resultRelations (i 1)
   :relationOids (o 16388 16388)
   ...
   }
```

Here is a planned statement from pg optimizer:
```log
{PLANNEDSTMT 
   :commandType 4 
   :queryId 0 
   ... 
   :planTree 
      {MODIFYTABLE 
      ...
      :resultRelations (i 1)
      ...
      :plans (
         {SEQSCAN 
         ...
         :plan_node_id 1 
         ...
         :scanrelid 1
         ...
         }
      )
      ...
      }
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16388 
      :requiredPerms 10 
      ...
      }
   )
   :resultRelations (i 1)
   ...
   }
}
```

As a result after this patch there is only one RTE (deduplicated) (instead what was before), like it's done with pg planner. But, really, there is a problem: rte has only delete permissions. So i'll refactor `ProcessTalbeDescriptor` function

HustonMmmavr commented:
> A general question. What's the relation between queryID under the planned statement node and assigned_query_id introduced by this PR?

There is no relation between `queryID` from planned statement and `assigned_query_id` introduced by this PR. `assigned_query_id` is used only for detection of table descriptors which points to the same result RTE's of query structure at ModifyTable operations (to prevent duplicates of RTE at the final plan).

HustonMmmavr commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -536,6 +536,59 @@ CTranslatorDXLToPlStmt::SetParamIds(Plan *plan)
+		m_dxl_to_plstmt_context->AddRTE(rte);
+	}
+
+	*rtePtr = rte;
```
I think this is more suitable place, because the logic of appending rte to the `m_rtable_entries_list` (`AddRTE`) won't be duplicated between `TranslateDXLTblScan/TranslateDXLIndexScan/...` and other Scan funcs. But, seems there is a problem with `TranslateDXLTblDescrToRangeTblEntry` (this function [allocates](https://github.com/greenplum-db/gpdb/blob/06d89cf82f3348ebdfeab05e62c044441266ef27/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L4662) RangeTableEntry and it's [alias](https://github.com/greenplum-db/gpdb/blob/06d89cf82f3348ebdfeab05e62c044441266ef27/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L4678)), so in case of DML operation there would be two table descriptors which points to the same result relation, so there would be generated two rte's (by multiple call of TranslateDXLTblDescrToRangeTblEntry) and first of them would be saved at list - it's ok, but second will be only allocated, but not freed (also permissions in this case are not merged by bitwise OR)

THANATOSLAVA replied:
>I think this is more suitable place, because the logic of appending rte to the m_rtable_entries_list (AddRTE) won't be duplicated between TranslateDXLTblScan/TranslateDXLIndexScan/... and other Scan funcs.

Maybe I wasn't being very clear. I didn't mean we should keep using our current logic of appending `rte`. What I suggested was to change the implementation slightly for better readability. Currently, `ProcessTableDescr` does three things: (1) located index, (2) populate `rte` (passed in by reference), and (3) append `rte` when necessary. However, the main task of appending `rte` isn't reflected by the return type `Index`.

It would be more intuitive if we just use the function to determine `is_existed_rte` and locate `index`. Depending on `is_existed_rte`, we build and append `rte` in `TranslateDXLTblScan`.

THANATOSLAVA replied:
>But, seems there is a problem with TranslateDXLTblDescrToRangeTblEntry (this function allocates RangeTableEntry and it's alias), so in case of DML operation there would be two table descriptors which points to the same result relation, so there would be generated two rte's (by multiple call of TranslateDXLTblDescrToRangeTblEntry) and first of them would be saved at list - it's ok, but second will be only allocated, but not freed (also permissions in this case are not merged by bitwise OR)

Can this be avoided by some refactoring?

HustonMmmavr replied:
Yes, seems it's worth to move calculation of index into separate function (it can be moved inside `CContextDXLToPlStmt` class, so `m_dxl_to_plstmt_context` object will calculate an index and flag `is_existed_rte`). But, maybe it's worth, to move logic with `AddRTE` to `TranslateDXLTblDescrToRangeTblEntry`? (but I'm not sure that it will be more intuitive)
Current logic `TranslateDXLTblDescrToRangeTblEntry` is creating a new rte and filling `base_table_context` (`base_table_context` is needed by all Scan functions), so my proposal is to refactor `TranslateDXLTblDescrToRangeTblEntry`:
- move logic with AddRTE here: if `is_existed_rte==true` it will return an existing RTE (from `m_dxl_to_plstmt_context->m_rtable_entries_list`), else it will generate a new RTE and append it to list
- also logic with setting RTE permissions move to this function
- logic with filling `base_table_context` won't be refactored

HustonMmmavr commented:
What was done:
- rebased PR from main branch
- refactored `CTranslatorUtils::GetTableDescr()` [at this commit](3ffd8f104b2f156262d63abdf88bc012bf863a4d)
- refactored `CTranslatorDXLToPlStmt::ProcessTableDescr()` (it was removed) [at this commit](25542ed76421f09c339dd26b74505b68d219b9a3),
  also CTranslatorDXLToPlStmt::TranslateDXLAppend function was fixed (this PR firstly was written for 6.x and i've missed to refactor
  this function)
- fixed tests at the [at the last commit]() (update_gp - new test after rebase, catchache - test contains `Append` node)

THANATOSLAVA commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -3322,10 +3330,13 @@ CTranslatorQueryToDXL::TranslateRTEToDXLLogicalGet(const RangeTblEntry *rte,
 	}
 
+	BOOL is_result_rel = m_query->resultRelation > 0 &&
+						 ULONG(m_query->resultRelation) == rt_index;
```
Not saying it's wrong, but could you please explain this logic?

HustonMmmavr replied:
- `m_query->resultRelation` - is a rtable index of target relation `INSERT/UPDATE/DELETE`; 0 for SELECT ([from parsenodes.h](https://github.com/greenplum-db/gpdb/blob/dffad5f7aea4e71ea3a980f69eb45130f47da383/src/include/nodes/parsenodes.h#L150-L170)). This index points to rte at `m_query->rtable` list ([here is an example of extracting rte by `resultRelation` index](https://github.com/greenplum-db/gpdb/blob/dffad5f7aea4e71ea3a980f69eb45130f47da383/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L749-L750))
- `rt_index` - is an argument of `TranslateRTEToDXLLogicalGet` which is called from `TranslateFromClauseToDXL`.
TranslateFromClauseToDXL [extracts rtindex from the RangeTblRef node and by this value extracts rte from the `m_query->rtable` list](https://github.com/greenplum-db/gpdb/blob/dffad5f7aea4e71ea3a980f69eb45130f47da383/src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp#L3198-L3235) (here is also link to [RangeTblRef node](https://github.com/greenplum-db/gpdb/blob/dffad5f7aea4e71ea3a980f69eb45130f47da383/src/include/nodes/primnodes.h#L1592-L1603)).  So rt_tindex is the index of relation at  `m_query->rtable` list

To detect that current table descriptor points to the result relation of a query we need to check is user query is an `INSERT/UPDATE/DELETE` (`m_query->resultRelation > 0`) and check that index of result relation is the same with  `rte` which is currently translates (`ULONG(m_query->resultRelation) == rt_index`).

THANATOSLAVA replied:
Thanks!

HustonMmmavr commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -572,11 +568,11 @@ CTranslatorDXLToPlStmt::TranslateDXLTblScan(
-	GPOS_ASSERT(nullptr != rte);
-	rte->requiredPerms |= ACL_SELECT;
-	m_dxl_to_plstmt_context->AddRTE(rte);
+		dxl_table_descr, index, &base_table_context, is_rte_exist, ACL_SELECT);
```
No, rte is needed by code [below](https://github.com/greenplum-db/gpdb/blob/dffad5f7aea4e71ea3a980f69eb45130f47da383/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L600-L603)

THANATOSLAVA replied:
You are right

HustonMmmavr commented on src/backend/gpopt/translate/CTranslatorUtils.cpp:
```diff
@@ -137,8 +136,13 @@ CTranslatorUtils::GetTableDescr(CMemoryPool *mp, CMDAccessor *md_accessor,
+	// assigned_query_id field is used to tag relations assigned to target one.
+	// In case of possible nested DML subqueries this field points to target
+	// relation of corresponding Query structure of subquery.
+	ULONG assigned_query_id = is_result_rel ? query_id : UNASSIGNED_QUERYID;
```
Refactored at [this commit](https://github.com/greenplum-db/gpdb/pull/14304/commits/49f502b3704a274149fc11859ddbf48d5f4cda23)

THANATOSLAVA replied:
Thanks!

HustonMmmavr commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4714,23 +4682,44 @@ CTranslatorDXLToPlStmt::TranslateDXLTblDescrToRangeTblEntry(
-		(void) base_table_context->InsertMapping(dxl_col_descr->Id(), attno);
+	// descriptor was already processed, and translated RTE is stored at
+	// context rtable list (only update required perms of this rte is needed)
+	if (rte_was_translated)
```
Refactored at [this commit](https://github.com/greenplum-db/gpdb/pull/14304/commits/91ba181e56b3c295c022aa31e74fbae3a24bd934)

THANATOSLAVA replied:
?

What I suggested was moving this portion to the beginning of the function `TranslateDXLTblDescrToRangeTblEntry` and keep the rest like what we have in greenplum-db:main. Makes sense?

```
	// descriptor was already processed, and translated RTE is stored at
	// context rtable list (only update required perms of this rte is needed)
	if (rte_was_translated)
	{
		RangeTblEntry *rte = m_dxl_to_plstmt_context->GetRTEByIndex(index);
		GPOS_ASSERT(nullptr != rte);
		rte->requiredPerms |= acl_mode;
		return rte;
	}
```

HustonMmmavr replied:
There is  a problem, this function previously (and currently) makes two things:
1. fills `rte` and returns it
2. fills `base_table_context` (which is allocated on stack at callers: [TranslateDXLTblScan](https://github.com/greenplum-db/gpdb/blob/cb8d53187010fcfe8cc8c1ba649b7c0e50e614bd/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L586), [TranslateDXLIndexScan](https://github.com/greenplum-db/gpdb/blob/cb8d53187010fcfe8cc8c1ba649b7c0e50e614bd/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L748), and others)
This table context is needed for further transformation (for ex. [at TranslateDXLTblScan](https://github.com/greenplum-db/gpdb/blob/cb8d53187010fcfe8cc8c1ba649b7c0e50e614bd/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L615-L619) -> [TranslateProjListAndFilter](https://github.com/greenplum-db/gpdb/blob/cb8d53187010fcfe8cc8c1ba649b7c0e50e614bd/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L5156-L5166) -> [TranslateDXLProjList](https://github.com/greenplum-db/gpdb/blob/cb8d53187010fcfe8cc8c1ba649b7c0e50e614bd/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L4841-L4846)), so seems we can't largely keep  code base.

HustonMmmavr commented:
I've rebased patch again, due to updated tests from main branch, and here is some changes:
- refactored work with `rte_was_translated`, also extended commentary for `TranslateDXLTblDescrToRangeTblEntry` function at [this commit](https://github.com/greenplum-db/gpdb/pull/14304/commits/799d215cf52f770302c7790717a48174ef3834b5)
- moved calculation of query_id to caller at [this commit](https://github.com/greenplum-db/gpdb/pull/14304/commits/5cdc228614fbe279baa05ec9235e7e5481184d9e)

THANATOSLAVA commented on src/backend/gpopt/translate/CContextDXLToPlStmt.cpp:
```diff
@@ -567,4 +569,50 @@ CContextDXLToPlStmt::FindRTE(Oid reloid)
+							   GPOS_NEW(m_mp) Index(new_index));
+
+	return new_index;
+}
```
Thanks for the detailed explanation. Can we add those those two bullets as comments, including the example.

Also, what about the third situation, `assigned_query_id != 0` but `usedIndex == 0`?

```
	Index new_index = gpdb::ListLength(m_rtable_entries_list) + 1;
	m_used_rte_indexes->Insert(GPOS_NEW(m_mp) ULONG(assigned_query_id),
							   GPOS_NEW(m_mp) Index(new_index));
```

HustonMmmavr replied:
Sorry, forgot to mention about this.
It's a next case: result relation in case of `ModifyTable` operations has more than one table descriptors (but these table descriptors has their `tag` - `assigned _query_id` which shows that they are both points to the same result relation), so   table descriptor which process first (with help of it's field `assigned_query_id`) would create a new unique index for the result relation . Thus next table_descriptor, pointing to the same result relation, instead of creating new index (which leads to creating new range table entry and pushing it's duplicate to `m_dxl_to_plstmt_context->m_rtable_entries_list`) would use already processed.

THANATOSLAVA replied:
Hmmm I need a little more help here. First, lemme make sure I understood you correctly. Are you suggesting:

If, two different `assigned_query_id` point to the same result relation, such as
```
assigned_query_id_1 --> result_relation_x
assigned_query_id_2 --> result_relation_x
```
and if we are currently processing `assigned_query_id_1`, we will match it to a new result relation of index y, so that the key-value pairs become
```
assigned_query_id_1 --> result_relation_y
assigned_query_id_2 --> result_relation_x
```

If my understanding is correct, please see my question below.

I would imagine even if two different `assigned_query_id` point to the same result table, `Index *usedIndex = m_used_rte_indexes->Find(&assigned_query_id)` should still not return index 0.

`m_used_rte_indexes` is defined as a hash map using `assigned_query_id` as key, and rte index as value. Its `Find` function should return the rte index, instead of 0, right? Generally speaking, hash map doesn't reject duplicate values.

HustonMmmavr replied:
I think, `assigned_query_id` is a bad name (seems more comfortable `assigned_query_id_for_target_rel`), i'll refactor it.
`assigned_query_id` is a tag of table descriptors, it shows id of query structure which contains result relation. If table descriptors have the same `assigned_query_id` - these table descriptors point to the same result relation of `ModifyTable` operation.
So, at the final stage (transformation from `dxl` to `PlStmt`) we will have "duplicated" table descriptors (in case of `ModifyTable` operation) which points to the same result relation, and if some of these table descriptors have the same value of `assigned_query_id` - these descriptors point to the same result relation. 
```
assigned_query_id_1 --> result_relation_1
assigned_query_id_2 --> result_relation_2

if assigned_query_id_1 == assigned_query_id_2 then result_relation_1 is the result_relation_2
```

Thus if `assigned_query_id` of table descriptor wasn't processed already - the new result relation is found (new index created and so on), other table descriptors (which would be processed a bit later) with the same value of `assigned_query_id` won't create a new index (because this result relation was already processed).

THANATOSLAVA replied:
Gotcha. For the record I will summarize the idea here:

`m_used_rte_indexes` is a hash map, mapping `assigned_query_id` (shortened as "id") to the rte index (shortened as "index") as in `m_rtable_entries_list` (shortened as "list"). The hash map only stores positive id's, because id 0 is reserved for operations that are not INSERT/UPDATE/DELETE.

To look up the index, we may encounter three scenarios:

1. If `id == 0`, it indicates the operation is not INSERT/UPDATE/DELECT, and doesn't have a result relation. Since the index isn't stored in the hash map, we return `list_length+1`. This means we will add a new rte to the rte list.
2. If `id > 0`, it indicates the operation is INSERT/UPDATE/DELECT. We look for its index in the hash map. If the index is found, we return the index. This means we can reuse the rte that's already in the list.
3. If `id > 0`, but the index isn't found in the hash map, it means the id hasn't been processed. We return `list_length+1` as the index. This means we will add a new rte to the rte list. At the same time, we add the key-value pair `id --> list_length+1` to the hash map for future look ups.

HustonMmmavr replied:
I would add to the first paragraph, that `id`=0 also exists in insert/update/delete queries only at table descriptors that doesn't points to result relation

HustonMmmavr commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -3339,10 +3344,20 @@ CTranslatorQueryToDXL::TranslateRTEToDXLLogicalGet(const RangeTblEntry *rte,
+	// table_desc_query_id is used to tag table descriptors assigned to target
+	// (result) relations one. In case of possible nested DML subqueries it's
+	// field points to target relation of corresponding Query structure of subquery.
+	BOOL table_desc_query_id = UNASSIGNED_QUERYID;
```
not bool

THANATOSLAVA commented on src/backend/gpopt/translate/CContextDXLToPlStmt.cpp:
```diff
@@ -567,4 +569,80 @@ CContextDXLToPlStmt::FindRTE(Oid reloid)
+	Index *usedIndex =
+		m_used_rte_indexes->Find(&assigned_query_id_for_target_rel);
+
+	//	`usedIndex` is a non zero value in next case: table descriptor with
```
Is `usedIndex` non zero or non NULL?

HustonMmmavr replied:
Sorry, I've made a typo. It's NULL

HustonMmmavr replied:
Fixed

HustonMmmavr commented:
@THANATOSLAVA thank you for the review!
I've fixed the last problem with commentary. Does any additional changes is needed (squashing of commits, rebasing) to merge this patch?

chrishajas commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4798,6 +4792,8 @@ CTranslatorDXLToPlStmt::TranslateDXLTblDescrToRangeTblEntry(
 
+	m_dxl_to_plstmt_context->AddRTE(rte);
+
 	return rte;
```
Should we still be returning the rte here?

HustonMmmavr replied:
For current implementation, returning of `rte` is needed only for [TranslateDXLTblScan](https://github.com/greenplum-db/gpdb/blob/ca28a0c00e02d91c43f5f901759e1589fb842a53/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L602-L630). But I understood, it could be avoided, inserting next code inside [TranslateDXLTblScan if statement](https://github.com/greenplum-db/gpdb/blob/ca28a0c00e02d91c43f5f901759e1589fb842a53/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L624-L630). 
```c
RangeTblEntry *rte = m_dxl_to_plstmt_context->GetRTEByIndex(index);
```

And there is two possible solutions:
- simply change return value  of `TranslateDXLTblDescrToRangeTblEntry` from `RangeTblEntry*` to `void`
- move duplicated [code with index calculation](https://github.com/greenplum-db/gpdb/blob/ca28a0c00e02d91c43f5f901759e1589fb842a53/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L599-L602) inside `TranslateDXLTblDescrToRangeTblEntry`. So small refactor of `TranslateDXLTblDescrToRangeTblEntry` is needed (change return type from `RangeTblEntry*` to `Index`, change it's arguments and maybe renaming)

For me, second approach seems more convenient

chrishajas replied:
Cool, yeah that looks good. Thanks!

chrishajas commented on src/backend/gporca/libnaucrates/src/operators/CDXLTableDescr.cpp:
```diff
@@ -218,6 +220,13 @@ CDXLTableDescr::SerializeToDXL(CXMLSerializer *xml_serializer) const
 
+	if (UNASSIGNED_QUERYID != m_assigned_query_id_for_target_rel)
+	{
+		xml_serializer->AddAttribute(
```
Can we also add mdp test that uses this new attribute

HustonMmmavr replied:
done

chrishajas reviewed (approved):
Overall looks really good! A couple minor things, and then we'll get this in.

THANATOSLAVA commented:
**Finalizing commit message (in process)**

**Problem description**
Currently EPQ mechanism doesn't work correctly in ORCA.

**Example**
```sql
set gp_enable_global_deadlock_detector = on 
create table test as select 0 as i distributed randomly;

-- Start transaction
-- First session increments the value from 0 to 1
1: begin;
1: update test set i = i + 1;

-- Second session attempts to delete the old value 0
2: delete from test where i = 0;
1: end;

-- The updated row ends up being deleted
-- This is not the expected behavior
```

Unlike postgres plan, `SEQSCAN` node (corresponding to `ModifyTable` node) doesn't points to the target(result) relation of the plan. `scanrelid` of `SEQSCAN` node points to the `RTE` `2`, but `resultRelation` has index `1`. Here is a plan:

```
{PLANNEDSTMT 
    ...
   :planTree 
      {MODIFYTABLE
      ... 
      :resultRelations (i 1)
      ...
      :plans (
         {SEQSCAN 
         ...
         :plan_node_id 1
         ... 
         :scanrelid 2
         ...
         }
      )
      ...
      }
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 8 
      ...
      }
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 2 
      ...
      }
   )
   :resultRelations (i 1)
   :relationOids (o 16548 16548 16548)
   ...
   }
```

EPQ mechanism through function call chain calls `ExecScanFetch`.
Due to incorrect `scanrelid` at `SEQSCAN` node next branch under  `if (estate->es_epqTupleSet[scanrelid - 1])` won't be executed. Thus instead of working with slot from `es_epqTuple` cache - slot with an old row is returned from access method routine - `(*accessMtd) (node)` (which uses snapshot for delete operation). As a result `EvalPlanQual` returns not null slot with an old version of row to `ExecDelete`, but [tupleid points to the updated version and deletion is performed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/executor/nodeModifyTable.c#L943-L957)

```c
static inline TupleTableSlot *
ExecScanFetch(ScanState *node,
			  ExecScanAccessMtd accessMtd,
			  ExecScanRecheckMtd recheckMtd)
{
	EState	   *estate = node->ps.state;
    //...
	if (estate->es_epqTupleSlot != NULL)
	{
        // ...
		Index		scanrelid = ((Scan *) node->ps.plan)->scanrelid;

        // ...
		else if (estate->es_epqTupleSlot[scanrelid - 1] != NULL)
		{
			TupleTableSlot *slot = node->ss_ScanTupleSlot;

			/* Return empty slot if we already returned a tuple */
			if (estate->es_epqScanDone[scanrelid - 1])
				return ExecClearTuple(slot);
			/* Else mark to remember that we shouldn't return more */
			estate->es_epqScanDone[scanrelid - 1] = true

			slot = estate->es_epqTupleSlot[scanrelid - 1];
			//...
			return slot;
		}
	}

	/*
	 * Run the node-type-specific access method function to get the next tuple
	 */
	return (*accessMtd) (node);
}
```

At the initial Query structure (parsed and analyzed query) we have right links to `RTE`'s, so to fix the problem with `scanrelid` - we want to transfer this information to the initial algebrized plan tree of orca. So we assign the `queryid` to each query structure (for subqueruies `queryid` assigned too) and, also, we assign `queryid` to table descriptors `CDXLTableDescr` which explicitly refers to target relation of DML (`ModifyTable`). After orca planning transformations this value exposes to which target relation of DML (`ModifyTable`) operation is referred the current table descriptor. And this allows to correctly assign `RTE` indexes in result plan under transformation from `DXL` representation.

## Forced changes

After solving problem with `scanrelid` for `*SCAN` (`SEQSCAN`, `INDEXSCAN` and others) nodes of DML (`ModifyTable`) node
(`scanrelid` of `*SCAN` node corresponding to `ModifyTable` node points to the result/target relation of operation) output of `EXPLAIN` has changed. Thus `regress` tests and `modify_table_data_corrupt` test from `isolation2` was changed.
These changes could be divided into two groups:

### Tests that contains one table at DML (`ModifyTable`) operation

Here is exapmle of some test cases (links from master):

* [updatable_views_optimizer - prefixes at coulumn names and table alias are not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/updatable_views_optimizer.out#L404-L416)

* [update_gp_optimizer - prefixes at coulumn names and table alias are not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/update_gp_optimizer.out#L435-L443)

* [bfv_dml_optimizer - alias for table name is not needed](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/bfv_dml_optimizer.out#L185-L189)

At these tests DML (`ModifyTable`) operation is performed on the one table. Due to fixes of this patch, `ModifyTable` operations at these tests has only one (single) RTE - target relation, unlike it was before (scanrelid of `*SCAN` plan node (correspoinmg to `ModifyTalbe`) didn't point to the result (target) relation and plan contained two rtables, but there should be only one). Thus prefixes to column names at queries, that uses one table is not shown, like it is in postgres. Here is some links to code, related to showing prefix ([setting flag to show prefix](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain_gp.c#L2109) and and function `get_variable`, where prefix [may be appended](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/utils/adt/ruleutils.c#L6852)).
Also at some tests `explain` aliases was [deleted, because them are equal to the original relation name](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L4033-L4035)

### Tests with explain (used more than one relation)

Here is an example of some test case (links from master):

* [qp_subquery_optimizer - aliases for table name in explain shouldn't have "_1" suffix](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/test/regress/expected/qp_subquery_optimizer.out#L1446-L1453)

This test group was changed due to fixes of `scanrelid` too. `scanrelid` for `*SCAN` node of corresponoding `ModifyTable` node now points to the target(result) relation (and rtable list doesn't contain duplicate of `ModifyTable` target relation). Thus `ExplainPrintPlan` calculates list of `relation` names for explain a bit different (in comparsion to what was before) (calcultion is [based on "rels_used" bitmap](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L895-L897), which now [calculates correctly](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L1338-L1383) - `ModifyTable` node doesn't insert at bitmap by the fact unused rti, like it was before, so there is no phantom relation name in `rtable_names`). So position of table names (`explain` aliases) with suffixes changed due to fixes of `scanrelid`'. Like at previous test group at some tests `explain` aliases was [deleted, because them are equal to the original relation name](https://github.com/greenplum-db/gpdb/blob/84a251adfdd87ad8d39af4d123083a7869a21742/src/backend/commands/explain.c#L4033-L4035)

### Tests for the patch

The `gdd/concurrent_update.sql` (`expected/gdd/concurrent_update_optimizer.out`) was partially [ported from the 6X_STABLE](https://github.com/greenplum-db/gpdb/blob/f1bf39d23057f24aac1189577262340e4b72054b/src/test/isolation2/sql/gdd/concurrent_update.sql) branch and extended with test case from the description.

THANATOSLAVA commented:
**Final commit message**

**Issue**
Currently EPQ mechanism doesn't work correctly in ORCA.

**Example**
```sql
set gp_enable_global_deadlock_detector = on 
create table test as select 0 as i distributed randomly;

-- Start transaction
-- First session increments the value from 0 to 1
1: begin;
1: update test set i = i + 1;

-- Second session attempts to delete the old value 0
2: delete from test where i = 0;
1: end;

-- The updated row ends up being deleted
-- This is not the expected behavior
```

**Root cause**
As shown in the following plan tree, `MODIFYTABLE->resultRelations` indicates rte 1 is the result relation. However, `SEQSCAN->scanrelid` points to rte 2. And this causes the wrong slot (in this case the cache slot with the old value) to be handed for deletion operation. This is because the correct rte links aren't transferred to the algebrized query (logical expression). 

```
{PLANNEDSTMT 
    ...
   :planTree 
      {MODIFYTABLE
      ... 
      :resultRelations (i 1)
      ...
      :plans (
         {SEQSCAN 
         ...
         :plan_node_id 1
         ... 
         :scanrelid 2
         ...
         }
      )
      ...
      }
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 8 
      ...
      }
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname test 
         :colnames (""i"")
         }
      :relid 16548 
      :requiredPerms 2 
      ...
      }
   )
   :resultRelations (i 1)
   :relationOids (o 16548 16548 16548)
   ...
   }
```

**Solution**
We assign a query id to each query structure, including subqueries. The query id would be attached to the table descriptors, and is used to direct us to the target relation of DML. We maintain a `query id --> rte index` map. A new rte would be added only if it's not already in the map. An old rte would be reused, potentially having its permission updated. The idea is to allow multiple query id's being directed to the same rte. With this patch, the rte's in the plan tree won't have duplicates.

HustonMmmavr commented:
@THANATOSLAVA, @chrishajas Thanks!
